README.md
# CLI File Reader Library
A .NET library developed for the CodeBricks research project that is
specifically designed to read and rewrite .NET binaries and was used
in conjuction with the Visual Storms / Sharp Storms projects.

Original Project Website – https://archive.codeplex.com/?p=clifilerw


Copyright © 2002,2003 Antonio Cisternino (cisterni@di.unipi.it)

This library has been implemented  to provide access to the raw format
of CLI files. The library is  still under development, so you can find
some problem in its use.

The  library  has been  developed  using  CLI standard  documentation.
Because fast access  to files is a goal I use  memory mapping in order
to have fast access to the  file. This implies that you need a certain
amount of privileges to use it.

The  library just  exposes  the content  of  the CLI  file. There  are
facilities   to    help   the   programmer    to   interoperate   with
System.Reflection  but these  facilities  are distinct  from the  base
functionality. For instance whenever a string is required its index is
returned.  The programmer  must explictly  use the  Strings  object in
order to get a string.


ChangeLog
2004-09-05    <cisterni@CHIP>

	* CLIFile.cs, Enums.cs, Cursors.cs: Started the support for CLR v. 2.0
	assemblies. We have introduced the three new metadata tables (GenericParam,
	MethodSpec, and GenericParamConstraint). Still the bridge between data and
	metadata should be crossed, as well as checking the header.

2004-02-29    <cisterni@CHIP>

	* CLIFile.cs: Fixed a bug, now the label is preserved correctly between save
	and restore of a cursor.

	* OPCODES.cs: Added the list of opcodes of the runtime.
	
	* CLIFile.cs: Added the Normalize method to the ILInstruction struct to
	simplify the analisys of IL stream.

2004-02-26    <cisterni@CHIP>

	* CLIFile.cs: Major change! Move to 1.1 version. Introduced CLIFileReader
	objects caching using WeakReferences. Factory pattern is used to ensure
	caching.
	
	* CLIFile.cs: Introduced the FindMethod static method to simplify obtaining
	the method body given the MethodInfo reflection object.
	
2004-02-25    <cisterni@CHIP>

	* Cursors.cs: Added the FindToken method to MethodTableCursor to find the
	index of a method given the MethodInfo. This is useful to quickly locate
	a method within the assembly file.
	
	* CLIFile.cs: Added the possibility of acquiring the state of an ILCursor
	and restore it later through a static method. This change has required
	several changes in order to pass around the method index to the ILCursor.
	
	* Cursors.cs: MethodTableCursor changed to pass the index of the method
	table into the method body and subsequently into the ILCursor.

2003-04-25    <cisterni@CHIP>

	* CLIFile.cs: Added the cliHd private member that refers to the CLI
	file header. Added the EntryPointToken read only property that returns
	the token of the entry point.

2003-03-21    <cisterni@CHIP>

	* CLIFile.cs: ShortInlineI was erroneously treated as 2 bytes value.
	Now it is treated as a signed byte.

2003-03-12    <cisterni@CHIP>

	* CLIFile.cs (ILCursor): Added TargetsCount property to check how
	many targets have been found by TrackTargets().

2003-03-08    <cisterni@CHIP>

	* CLIFile.cs (SignatureUtil.FindType): Added a parameter to find
	types passed by ref. TypeArray and FindType has been also modified.

2003-03-06    <cisterni@CHIP>

	* CLIFile.cs (ILCursor): Added the private struct State to save and
	restore the state of the cursor. Added the saved member which is a
	Stack of States. PushState and PopState allow to save and restore the
	state of the cursor. This feature is useful for lookahead on the code.

2003-03-04    <cisterni@CHIP>

	* CLIFile.cs: Fixed a bug in IlineSwitch handling. now if TrackTargets
	is avoided the sw array is returned.

2003-03-03    <cisterni@CHIP>

	* CLIFile.cs: Added support for InlineSwitch operand type in
	TrackTargets.
	ResolveParameter has been changed to take into account also
	switches.

2003-02-24    <cisterni@CHIP>

	* CLIFile.cs: Added the path and assembly fields to CLIFileReader.
	These members are intended to address a problem in reflection.
	When assemblies are loaded using Assembly.LoadFrom Type.GetType
	returns false. We use GetType from the assembly object stored in
	assembly. To avoid unnecessary overhead the assembly field is
	instantiated only if GetType is invoked.

	* Cursors.cs: GetReflectionType() in TypeDefTableCursor has been
	changed in order to use table.CLIFile.assembly.GetType() instead
	of Type.GetType(). In this way the method works also with dynamically
	loaded assemblies.

2003-01-18    <cisterni@CHIP>

	* CLIFile.cs: ShortInlineI was considered of 2 bytes. It is now
	signed byte.

2003-01-16    <cisterni@CHIP>

	* CLIFile.cs: Bugfix in TrackTarget resolution.

	* CLIFile.cs: Bugfix in Next(), the array of labels was accessed
	past the end.

	* CLIFile.cs: Bugfix, when the branch of a target was 0 a wrong
	target position was generated.

2003-01-04    <cisterni@CHIP>

	* CLIFile.cs: TrackTargets now avoid to recalculate targets.

2002-12-19    <cisterni@CHIP>

	* CLIFile.cs: Now the par field of Instr struct is a struct with
	an explicit layout that behaves as a C union to save memory.

	* Enum.cs: Now this file is generated by OpCodesGen using Enum.cs.in as
	input file.

2002-12-18    <cisterni@CHIP>

	* CLIFile.cs: New change in the architecture of ILCursor. The cursor
	now has most of the code. There is the interface ILStoreCursor that
	abstracts the notion of cursor to the stream. This can be a mapptr or
	a cursor to an array of bytes. We pay some virtual call because of the 
	interface but it seems to be acceptable with respect to the previous
	solution. Even Generics wouldn't improve the situation.

	* CLIFile.cs: A new class ILCursor has been introduced. The ILCursor
	class is abstract and serves as base class for MapPtrILCursor and
	ByteArrayILCursor. The pair ILCursor and MapPtrILCursor corresponds to
	the previous ILCursor class. ByteArrayILCursor run across a byte array
	rather than memory and inherits from ILCursor.
	In this way we have introduced a number of virtual methods though we 
	don't believe that this would introduce any penalty in execution speed.

2002-10-21    <cisterni@CHIP>

	* Cursors.cs (SEHTableCursor): Added the Rows property that returns
	the number of rows of SEH table. The property lies here because there
	is no SEHTable. In future we may want to introduce it, though it may
	introduce unneeded overhead.

	* CLIFile.cs (SignatureUtil): The class has become public with the
	method SetTypeDefTable public and all other members internal. This
	is due to the fact that users of the library would be interested in
	setting the TypeDefTable.

	* Enums.cs (HasCustomattributeTag): Cut and past error. The 
	AssemblyRef constant was misdefined. (by Donal Lafferty)

	* CLIFile.cs (ResolveParameter): Fixed a bitmask (by Donal Lafferty)

2002-10-14    <cisterni@CHIP>

	* CLIFile.cs (SignatureUtil): Added manualTypeDefObjects and
	typeDefTable to control the FindType helper. (by Donal Lafferty).

OPCODES.cs
/// ------------------------------------------------------------
/// Copyright (c) 2002-2008 Antonio Cisternino (cisterni@di.unipi.it)
/// 
/// File: OPCODES.cs
/// Author: Antonio Cisternino (cisterni@di.unipi.it)
/// 
/// The use and distribution terms for this software are 
/// contained in the file named license.txt, which can be found 
/// in the root of this distribution.
/// By using this software in any fashion, you are agreeing to 
/// be bound by the terms of this license.
///
/// You must not remove this notice, or any other, from this
/// software.
/// ------------------------------------------------------------

using System;

namespace CLIFileRW {
  public enum OPCODES {
    Nop = 0,
    Break = 1,
    Ldarg_0 = 2,
    Ldarg_1 = 3,
    Ldarg_2 = 4,
    Ldarg_3 = 5,
    Ldloc_0 = 6,
    Ldloc_1 = 7,
    Ldloc_2 = 8,
    Ldloc_3 = 9,
    Stloc_0 = 10,
    Stloc_1 = 11,
    Stloc_2 = 12,
    Stloc_3 = 13,
    Ldarg_S = 14,
    Ldarga_S = 15,
    Starg_S = 16,
    Ldloc_S = 17,
    Ldloca_S = 18,
    Stloc_S = 19,
    Ldnull = 20,
    Ldc_I4_M1 = 21,
    Ldc_I4_0 = 22,
    Ldc_I4_1 = 23,
    Ldc_I4_2 = 24,
    Ldc_I4_3 = 25,
    Ldc_I4_4 = 26,
    Ldc_I4_5 = 27,
    Ldc_I4_6 = 28,
    Ldc_I4_7 = 29,
    Ldc_I4_8 = 30,
    Ldc_I4_S = 31,
    Ldc_I4 = 32,
    Ldc_I8 = 33,
    Ldc_R4 = 34,
    Ldc_R8 = 35,
    Dup = 37,
    Pop = 38,
    Jmp = 39,
    Call = 40,
    Calli = 41,
    Ret = 42,
    Br_S = 43,
    Brfalse_S = 44,
    Brtrue_S = 45,
    Beq_S = 46,
    Bge_S = 47,
    Bgt_S = 48,
    Ble_S = 49,
    Blt_S = 50,
    Bne_Un_S = 51,
    Bge_Un_S = 52,
    Bgt_Un_S = 53,
    Ble_Un_S = 54,
    Blt_Un_S = 55,
    Br = 56,
    Brfalse = 57,
    Brtrue = 58,
    Beq = 59,
    Bge = 60,
    Bgt = 61,
    Ble = 62,
    Blt = 63,
    Bne_Un = 64,
    Bge_Un = 65,
    Bgt_Un = 66,
    Ble_Un = 67,
    Blt_Un = 68,
    Switch = 69,
    Ldind_I1 = 70,
    Ldind_U1 = 71,
    Ldind_I2 = 72,
    Ldind_U2 = 73,
    Ldind_I4 = 74,
    Ldind_U4 = 75,
    Ldind_I8 = 76,
    Ldind_I = 77,
    Ldind_R4 = 78,
    Ldind_R8 = 79,
    Ldind_Ref = 80,
    Stind_Ref = 81,
    Stind_I1 = 82,
    Stind_I2 = 83,
    Stind_I4 = 84,
    Stind_I8 = 85,
    Stind_R4 = 86,
    Stind_R8 = 87,
    Add = 88,
    Sub = 89,
    Mul = 90,
    Div = 91,
    Div_Un = 92,
    Rem = 93,
    Rem_Un = 94,
    And = 95,
    Or = 96,
    Xor = 97,
    Shl = 98,
    Shr = 99,
    Shr_Un = 100,
    Neg = 101,
    Not = 102,
    Conv_I1 = 103,
    Conv_I2 = 104,
    Conv_I4 = 105,
    Conv_I8 = 106,
    Conv_R4 = 107,
    Conv_R8 = 108,
    Conv_U4 = 109,
    Conv_U8 = 110,
    Callvirt = 111,
    Cpobj = 112,
    Ldobj = 113,
    Ldstr = 114,
    Newobj = 115,
    Castclass = 116,
    Isinst = 117,
    Conv_R_Un = 118,
    Unbox = 121,
    Throw = 122,
    Ldfld = 123,
    Ldflda = 124,
    Stfld = 125,
    Ldsfld = 126,
    Ldsflda = 127,
    Stsfld = 128,
    Stobj = 129,
    Conv_Ovf_I1_Un = 130,
    Conv_Ovf_I2_Un = 131,
    Conv_Ovf_I4_Un = 132,
    Conv_Ovf_I8_Un = 133,
    Conv_Ovf_U1_Un = 134,
    Conv_Ovf_U2_Un = 135,
    Conv_Ovf_U4_Un = 136,
    Conv_Ovf_U8_Un = 137,
    Conv_Ovf_I_Un = 138,
    Conv_Ovf_U_Un = 139,
    Box = 140,
    Newarr = 141,
    Ldlen = 142,
    Ldelema = 143,
    Ldelem_I1 = 144,
    Ldelem_U1 = 145,
    Ldelem_I2 = 146,
    Ldelem_U2 = 147,
    Ldelem_I4 = 148,
    Ldelem_U4 = 149,
    Ldelem_I8 = 150,
    Ldelem_I = 151,
    Ldelem_R4 = 152,
    Ldelem_R8 = 153,
    Ldelem_Ref = 154,
    Stelem_I = 155,
    Stelem_I1 = 156,
    Stelem_I2 = 157,
    Stelem_I4 = 158,
    Stelem_I8 = 159,
    Stelem_R4 = 160,
    Stelem_R8 = 161,
    Stelem_Ref = 162,
    Ldelem = 163,
    Stelem = 164,
    Unbox_Any = 165,
    Conv_Ovf_I1 = 179,
    Conv_Ovf_U1 = 180,
    Conv_Ovf_I2 = 181,
    Conv_Ovf_U2 = 182,
    Conv_Ovf_I4 = 183,
    Conv_Ovf_U4 = 184,
    Conv_Ovf_I8 = 185,
    Conv_Ovf_U8 = 186,
    Refanyval = 194,
    Ckfinite = 195,
    Mkrefany = 198,
    Ldtoken = 208,
    Conv_U2 = 209,
    Conv_U1 = 210,
    Conv_I = 211,
    Conv_Ovf_I = 212,
    Conv_Ovf_U = 213,
    Add_Ovf = 214,
    Add_Ovf_Un = 215,
    Mul_Ovf = 216,
    Mul_Ovf_Un = 217,
    Sub_Ovf = 218,
    Sub_Ovf_Un = 219,
    Endfinally = 220,
    Leave = 221,
    Leave_S = 222,
    Stind_I = 223,
    Conv_U = 224,
    Prefix7 = 248,
    Prefix6 = 249,
    Prefix5 = 250,
    Prefix4 = 251,
    Prefix3 = 252,
    Prefix2 = 253,
    Prefix1 = 254,
    Prefixref = 255,
    Arglist = 256 + 0,
    Ceq = 256 + 1,
    Cgt = 256 + 2,
    Cgt_Un = 256 + 3,
    Clt = 256 + 4,
    Clt_Un = 256 + 5,
    Ldftn = 256 + 6,
    Ldvirtftn = 256 + 7,
    Ldarg = 256 + 9,
    Ldarga = 256 + 10,
    Starg = 256 + 11,
    Ldloc = 256 + 12,
    Ldloca = 256 + 13,
    Stloc = 256 + 14,
    Localloc = 256 + 15,
    Endfilter = 256 + 17,
    Unaligned = 256 + 18,
    Volatile = 256 + 19,
    Tailcall = 256 + 20,
    Initobj = 256 + 21,
    Constrained = 256 + 22,
    Cpblk = 256 + 23,
    Initblk = 256 + 24,
    Rethrow = 256 + 26,
    Sizeof = 256 + 28,
    Refanytype = 256 + 29,
    Readonly = 256 + 30
  }
}
